[gn, crashpad] Add GN build support and crashpad port

vejmartin opened the pull request:
Adds support for `depot_tools` through `vcpkg_acquire_depot_tools`, as well as a port of [Crashpad](https://chromium.googlesource.com/crashpad/crashpad)

- What does your PR fix? Fixes issue: https://github.com/microsoft/vcpkg/issues/9650
Follow-up to: https://github.com/microsoft/vcpkg/pull/10330 and https://github.com/microsoft/vcpkg/pull/10083#issuecomment-586666639

- Which triplets are supported/not supported? 
`x64-osx`, `x64-windows-static` (`x64-windows` switches to static linking)

- Have you updated the CI baseline?
Yes

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

Notes: 
* `vcpkg_acquire_depot_tools` (name consistent with `vcpkg_acquire_msys`) finds or acquires `depot_tools` and also supports finding `gclient` and `gn` through the `TOOL` parameter.
*  The `gclient` and `gn` tools work by directly calling the python scripts, rather than the shell wrappers. The shell wrappers for `gclient` include the logic for depot_tools' auto-update and bootstrap functionalities which are unnecessary as `vcpkg` already includes `python` and `git`.
* `crashpad` is fetched directly with `gclient` instead of with the `fetch` tool, so that a commit hash may be specified through the `.gclient` config.
* `crashpad` has a dependency on `buildtools`, which could be disabled once `gn` becomes available as part of `vcpkg_find_acquire_program`. (https://github.com/microsoft/vcpkg/pull/10330)

vejmartin committed:
[crashpad] Add new port

JackBoosY commented:
Another related PR: #8963.

JackBoosY commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,144 @@
+    ADD_TO_PATH)
+
+message(STATUS "Syncing sources and dependencies...")
+vcpkg_execute_required_process(
```
Can we take this step out into a function?
Such as _VCPKG_PATH/scripts/vcpkg_sync_gn.cmake_.

vejmartin replied:
This has now been moved to `vcpkg_from_gclient`, as it's possible to use `gclient` without `gn`.

JackBoosY commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,144 @@
+    set(BUILD_DIR "${CURRENT_BUILDTREES_DIR}/${CONFIG}")
+
+    message(STATUS "Generating build (${CONFIG})...")
+    vcpkg_execute_required_process(
```
Can we take this step out into a function?
Such as _VCPKG_PATH/scripts/vcpkg_configure_gn.cmake_.

vejmartin replied:
Yup, moved to `vcpkg_configure_gn.cmake`.

JackBoosY commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,144 @@
+    set(BUILD_DIR "${CURRENT_BUILDTREES_DIR}/${CONFIG}")
+
+    message(STATUS "Building (${CONFIG})...")
+    vcpkg_execute_build_process(
```
Can we take this step out into a function?
Such as _VCPKG_PATH/scripts/vcpkg_build_gn.cmake_.

vejmartin replied:
Moved this to `vcpkg_build_ninja.cmake`, as `GN` is only used to generate the ninja targets.

JackBoosY commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,144 @@
+    )
+endfunction()
+
+function(install_binaries CONFIG OUT_DIR_PREFIX)
```
Can we take this step out into a function?
Such as _VCPKG_PATH/scripts/vcpkg_install_gn.cmake_.

vejmartin replied:
I've split this into `vcpkg_install_library.cmake` and `vcpkg_install_tool.cmake` to make it easier to be used by other ports as well.

JackBoosY commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,144 @@
+
+if(CMAKE_HOST_WIN32)
+    # Load toolchains
+    if(NOT VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
```
These settings need to be uniformly set in a function.

vejmartin replied:
Unfortunately, the formatting of the options here is entirely dependent on the `GN` project. To illustrate, Chromium uses `extra_cppflags=string` vs Crashpad's `extra_cflags_cc=string` vs Skia's `extra_cflags_cc=[list of strings]`

JackBoosY commented on scripts/cmake/vcpkg_from_git.cmake:
```diff
@@ -77,13 +97,13 @@ function(vcpkg_from_git)
       COMMAND ${GIT} fetch ${_vdud_URL} ${_vdud_REF} --depth 1 -n
       WORKING_DIRECTORY ${DOWNLOADS}/git-tmp
-      LOGNAME git-fetch-${TARGET_TRIPLET}
+      LOGNAME git-fetch-${LOGNAME}
```
We also need to specify the log name based on the different triplets.

vejmartin replied:
`LOGNAME` defaults to `${TARGET_TRIPLET}` if the `NAME` parameter has not been set. `vcpkg_acquire_depot_tools` sets the `NAME` parameter to make a distinction between fetching port and tool repositories.

JackBoosY commented:
/azp run

JackBoosY commented:
/azp run

JackBoosY commented:
@Neumann-A @Voskrese  Could you please review this PR?

Thanks.

Neumann-A commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+
+    set(OPTIONS_REL "${OPTIONS_REL} \
+        extra_cflags_c=\"${CMAKE_C_FLAGS} ${CMAKE_C_FLAGS_RELEASE}\" \
+        extra_cflags_cc=\"${CMAKE_CXX_FLAGS} ${CMAKE_CXX_FLAGS_RELEASE}\"")
```
this logic belongs into `vcpkg_configure_gn()` and not the portfile

vejmartin replied:
As described above (https://github.com/microsoft/vcpkg/pull/10505#discussion_r397654007), 
> Unfortunately, the formatting of the options here is entirely dependent on the `GN` project. To illustrate, Chromium uses `extra_cppflags=string` vs Crashpad's `extra_cflags_cc=string` vs Skia's `extra_cflags_cc=[list of strings]`

The flags for `Crashpad` are defined as part of the [toolchain](https://chromium.googlesource.com/chromium/mini_chromium/+/refs/heads/master/build/BUILD.gn) and need to be strings. Whereas the flags for `Skia` are added as a [config("extra_flags")](https://skia.googlesource.com/skia/+/refs/heads/master/gn/BUILD.gn) and need to be passed as a list of strings. `Chromium` uses a different name for the extra C++ flags, `extra_cppflags` in its [gcc_toolchain](https://chromium.googlesource.com/chromium/src/+/14b686f408617c9953d665171f550ea0a9fb06b7/build/toolchain/gcc_toolchain.gni), which might not be available on Windows, and so does [V8](https://chromium.googlesource.com/v8/v8.git/+/refs/heads/master/tools/toolchain/BUILD.gn).

A [custom toolchain or build config](https://chromium.googlesource.com/chromium/src/tools/gn/+/48062805e19b4697c5fbd926dc649c78b6aaa138/docs/language.md#build-configuration) for `vcpkg` might be able to normalize all of this, provided that all `GN`-based projects share a common set of requirements.

Neumann-A commented on ports/crashpad/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+    OPTIONS_DEBUG "${OPTIONS_DBG}"
+    OPTIONS_RELEASE "${OPTIONS_REL}"
+)
+vcpkg_build_ninja(client crashpad_handler)
```
should probably be `vcpkg_install_gn()` just like `vcpkg_install_meson()` which just calls `ninja install`
additional `vcpkg_build_gn()` might be required just to keep it similar with the other buildsystem (cmake, meson, make, msbuild)

vejmartin replied:
`Crashpad` and `Skia` (and possibly `GN` in general) don't define an `install` target. That step is left for the user of the library. While it might be consistent to call it `vcpkg_install_gn`, it won't exactly be accurate, as it will only build the targets and not install them (building all targets is unnecessary in most cases, so there must be a way to specify them).

Neumann-A commented on scripts/cmake/vcpkg_from_gclient.cmake:
```diff
@@ -0,0 +1,73 @@
+        COMMAND "${GCLIENT}" sync --no-history --shallow
+        WORKING_DIRECTORY "${SOURCE_PATH}"
+        LOGNAME gclient-sync-${TARGET_TRIPLET}
+    )
```
Is using gclient really required? We don't want to checkout the latest commit and we probably also don't want to check out dependencies since those should be already installed as a different port within vcpkg.

vejmartin replied:
`vcpkg_from_gclient` requires a `REF` parameter, which it passes to the `.gclient` config, so that it doesn't download the latest commit.

Without `gclient`, we'd have to manually satisfy the requirements in the `DEPS` files. `Crashpad` needs its dependencies to be in the locations specified in the [DEPS](https://chromium.googlesource.com/crashpad/crashpad/+/refs/heads/master/DEPS) file for `GN` to successfully generate the build (e.g. `third_party/mini_chromium` is used inside the [BUILDCONFIG.gn](https://chromium.googlesource.com/crashpad/crashpad/+/refs/heads/master/build/BUILDCONFIG.gn) file.

vejmartin commented:
There are a few issues blocking this PR and other `GN` projects that need more consideration:
* There is no common way to pass extra compiler or linker flags to `GN`, as this is supported on a per-project basis. One solution might be to define a custom `BUILDCONFIG.gn` for `vcpkg`.
* Each project requires its own in-source dependencies as outlined in the `DEPS` file, which is usually processed by `gclient` (in Skia there's a `git-sync-deps.py` script). Building these dependencies could be disabled with a patch, however, a full checkout is necessary, as GN fails whenever a dependency cannot be found. Unfortunately, this conflicts with the goal of reusing libraries already built with `vcpkg` and may also lead to filename conflicts.
* GN doesn't handle library installation, and so each project has to manually be installed to the packages directory.

Resolving the above could be done by patching the build config or by porting it all to CMake. The latter would also render `depot_tools` unnecessary. However, either of the two would need keeping in sync with upstream when updating.

Voskrese commented:
in depot_tools there is a script converter on python, from gn to cmake project, at least in the project skia

Voskrese commented:
https://github.com/google/skia/blob/master/gn/gn_to_cmake.py

> CMake
> 
> We have added a GN-to-CMake translator mainly for use with IDEs that like CMake project descriptions. This is not meant for any purpose beyond development.
> 
> bin/gn gen out/config --ide=json --json-ide-script=../../gn/gn_to_cmake.py
>

vejmartin commented:
@Voskrese, thank you for suggesting the `gn_to_cmake.py` script! While it does look like it has been originally designed for IDE support, it does look quite promising.  

There a few issues to fix to make it work with `crashpad`, but in general the `GN` library types (static, shared, source_set, group, loadable_module) should be able to map onto the ones in `CMake` (static, shared, object, interface, module).  

Something to note for this approach is that a full dependency checkout would still be required (`gclient sync`), as the `gn_to_cmake.py` script works with the `json` output from `GN`.  After the conversion, the in-source third-party libraries would exist as part of the `LINK_LIBRARIES` property of relevant targets and will need to be replaced with ones from `vcpkg`.

JackBoosY commented:
Please continue to complete this PR, most google ports need to use depot tools, we are looking forward merge this PR!

vejmartin commented:
Thank you, @JackBoosY . I've been looking into adapting the `gn_to_cmake.py` script to use modern CMake. At the moment it flattens the GN targets and configs, so the resulting CMake commands are very repetitive. This is an issue as dependencies can't cleanly be removed if their properties have already been propagated and flattened. I've submitted some changes to GN, which should make extracting the necessary info easier.  

It would also be important to validate that the `gn_to_cmake.py` script works exactly like the original build, perhaps by comparing the `compile_commands.json` from the two.

Another approach I'm also considering is to patch the `BUILD.gn` files for third-party dependencies and forward them back to the installed vcpkg ones. This would probably be less error-prone than using the `gn_to_cmake.py` script.

vejmartin committed:
[vcpkg_find_acquire_program] Add gn

vejmartin commented:
I've decided to address the dependency issues by replacing the corresponding `BUILD.gn` files and by using `vcpkg_from_git` where the prior isn't sufficient. 

In the case of `crashpad`, its build configuration exists mostly inside `mini_chromium`, which is checked out via `vcpkg_from_git` and built in-source. The in-source dependency on `zlib` is substituted with the one built by `vcpkg` by replacing the `crashpad/third_party/zlib/BUILD.gn` file with `ports/crashpad/zlib.gn`.

This approach eliminates the need for `gclient` and so I've also removed `depot_tools` from the original changelist and added `GN` to `vcpkg_find_acquire_program` (https://github.com/microsoft/vcpkg/pull/10330). Where `depot_tools` are required to port other libraries, it would be necessary to inspect the `DEPS` file in the root directory to resolve any dependencies in a similar way.

JackBoosY commented:
@vejmartin Please ping me if this PR is ready for review.

vejmartin commented:
@JackBoosY, it's ready for review :)

JackBoosY commented on scripts/cmake/vcpkg_build_ninja.cmake:
```diff
@@ -0,0 +1,37 @@
+## # vcpkg_build_ninja
```
I always think that this function should be changed to `gn` related names to distinguish different configuration environments.

JackBoosY reviewed (dismissed):
There are still some questions, please check my previous comments.

vejmartin committed:
[GN] Add build system support

vejmartin commented on docs/maintainers/vcpkg_build_ninja.md:
```diff
@@ -0,0 +1,17 @@
+# vcpkg_build_ninja
```
I've added `vcpkg_build_gn` function that wraps the call to `vcpkg_build_ninja`.

vejmartin commented on docs/maintainers/vcpkg_install_library.md:
```diff
@@ -0,0 +1,21 @@
+# vcpkg_install_library
```
I've added a `vcpkg_install_gn` function that uses the `gn desc` command to query the target outputs and type and removed the `vcpkg_install_library` and `vcpkg_install_tool` functions.

strega-nil commented:
Alright, this looks great! Thanks @vejmartin :)

JackBoosY commented:
This small step of PR is a big step of Google ports.

NoFr1ends commented:
@vejmartin Any particular reason why it's not supported on Linux? 
As far as I know crashpad does support linux.

vejmartin commented:
@NoFr1ends, I haven't had the chance to verify that it works on Linux. If it does, then it'd be great to enable it :)

jason-ortiz commented:
> @vejmartin Any particular reason why it's not supported on Linux? As far as I know crashpad does support linux.

@NoFr1ends , check out my PR #23827
